File: packages/react-native/ReactCommon/react/renderer/graphics/ClipPath.cpp
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include "ClipPath.h"

#include <folly/dynamic.h>
#include <sstream>
#include <string>
#include <vector>

namespace facebook::react {

bool CircleShape::operator==(const CircleShape& other) const {
  return cx == other.cx && cy == other.cy && r == other.r;
}

#if RN_DEBUG_STRING_CONVERTIBLE
void CircleShape::toString(std::stringstream& ss) const {
  ss << "circle(" << r << "px";
  if (cx || cy) {
    ss << " at ";
    if (cx) {
      ss << cx->toString();
    }
    if (cy) {
      ss << " " << cy->toString();
    }
  }
}
#endif

#ifdef RN_SERIALIZABLE_STATE
folly::dynamic CircleShape::toDynamic() const {
  folly::dynamic result = folly::dynamic::object();
  result["r"] = r.toDynamic();
  if (cx) {
    result["cx"] = cx->toDynamic();
  }
  if (cy) {
    result["cy"] = cy->toDynamic();
  }
  return result;
}
#endif

bool EllipseShape::operator==(const EllipseShape& other) const {
  return cx == other.cx && cy == other.cy && rx == other.rx && ry == other.ry;
}

#if RN_DEBUG_STRING_CONVERTIBLE
void EllipseShape::toString(std::stringstream& ss) const {
  ss << "ellipse(" << rx << "px " << ry << "px";
  if (cx || cy) {
    ss << " at ";
    if (cx) {
      ss << cx->toString();
    }
    if (cy) {
      ss << " " << cy->toString();
    }
  }
}
#endif

#ifdef RN_SERIALIZABLE_STATE
folly::dynamic EllipseShape::toDynamic() const {
  folly::dynamic result = folly::dynamic::object();
  result["rx"] = r.toDynamic();
  result["ry"] = r.toDynamic();
  if (cx) {
    result["cx"] = cx->toDynamic();
  }
  if (cy) {
    result["cy"] = cy->toDynamic();
  }
  return result;
}
#endif

bool InsetShape::operator==(const InsetShape& other) const {
  return top == other.top && right == other.right && bottom == other.bottom &&
      left == other.left && borderRadius == other.borderRadius;
}

#if RN_DEBUG_STRING_CONVERTIBLE
void InsetShape::toString(std::stringstream& ss) const {
  ss << "inset(" << top.toString() << " " << right.toString() << " "
     << bottom.toString() << " " << left.toString();
  if (borderRadius) {
    ss << " round " << borderRadius->toString();
  }
  ss << ")";
}
#endif

#ifdef RN_SERIALIZABLE_STATE
folly::dynamic InsetShape::toDynamic() const {
  folly::dynamic result = folly::dynamic::object();
  result["top"] = top.toDynamic();
  result["right"] = right.toDynamic();
  result["bottom"] = bottom.toDynamic();
  result["left"] = left.toDynamic();
  return result;
}
#endif

bool PolygonShape::operator==(const PolygonShape& other) const {
  return points == other.points;
}

#if RN_DEBUG_STRING_CONVERTIBLE
void PolygonShape::toString(std::stringstream& ss) const {
  ss << "polygon(";
  for (size_t i = 0; i < points.size(); i++) {
    if (i > 0) {
      ss << ", ";
    }
    ss << points[i].first.toString() << " " << points[i].second.toString();
  }
  ss << ")";
}
#endif

#ifdef RN_SERIALIZABLE_STATE
folly::dynamic PolygonShape::toDynamic() const {
  folly::dynamic result = folly::dynamic::array();
  for (const auto& point : points) {
    folly::dynamic pointObj = folly::dynamic::object();
    pointObj["x"] = point.first.toDynamic();
    pointObj["y"] = point.second.toDynamic();
    result.push_back(pointObj);
  }
  return result;
}
#endif

// RectShape implementations
bool RectShape::operator==(const RectShape& other) const {
  return top == other.top && right == other.right && bottom == other.bottom &&
      left == other.left && borderRadius == other.borderRadius;
}

#if RN_DEBUG_STRING_CONVERTIBLE
void RectShape::toString(std::stringstream& ss) const {
  ss << "rect(" << top.toString() << " " << right.toString() << " "
     << bottom.toString() << " " << left.toString() << " ";
  if (borderRadius) {
    ss << "round " << borderRadius->toString();
  }
  ss << ")";
}
#endif

#ifdef RN_SERIALIZABLE_STATE
folly::dynamic RectShape::toDynamic() const {
  return folly::dynamic::object()("top", top.toDynamic())(
      "right", right.toDynamic())("bottom", bottom.toDynamic())(
      "left", left.toDynamic());
}
#endif

// XywhShape implementations
bool XywhShape::operator==(const XywhShape& other) const {
  return x == other.x && y == other.y && width == other.width &&
      height == other.height && borderRadius == other.borderRadius;
}

#if RN_DEBUG_STRING_CONVERTIBLE
void XywhShape::toString(std::stringstream& ss) const {
  ss << "xywh(" << x.toString() << " " << y.toString() << " "
     << width.toString() << " " << height.toString();
  if (borderRadius) {
    ss << " round " << borderRadius->toString();
  }
  ss << ")";
}
#endif

#ifdef RN_SERIALIZABLE_STATE
folly::dynamic XywhShape::toDynamic() const {
  return folly::dynamic::object()("x", x.toDynamic())("y", y.toDynamic())(
      "width", width.toDynamic())("height", height.toDynamic());
}
#endif

// PathShape implementations
bool PathShape::operator==(const PathShape& other) const {
  return pathData == other.pathData;
}

#if RN_DEBUG_STRING_CONVERTIBLE
void PathShape::toString(std::stringstream& ss) const {
  ss << "path(\"" << pathData << "\")";
}
#endif

#ifdef RN_SERIALIZABLE_STATE
folly::dynamic PathShape::toDynamic() const {
  return folly::dynamic::object()("pathData", pathData);
}
#endif

bool ClipPath::operator==(const ClipPath& other) const {
  return shape == other.shape && geometryBox == other.geometryBox;
}

#if RN_DEBUG_STRING_CONVERTIBLE
std::string ClipPath::toString() const {
  std::stringstream ss;

  if (shape) {
    std::visit([&](const auto& s) { s.toString(ss); }, *shape);
  }

  if (geometryBox) {
    if (shape) {
      ss << " ";
    }
    switch (*geometryBox) {
      case GeometryBox::MarginBox:
        ss << "margin-box";
        break;
      case GeometryBox::BorderBox:
        ss << "border-box";
        break;
      case GeometryBox::ContentBox:
        ss << "content-box";
        break;
      case GeometryBox::PaddingBox:
        ss << "padding-box";
        break;
      case GeometryBox::FillBox:
        ss << "fill-box";
        break;
      case GeometryBox::StrokeBox:
        ss << "stroke-box";
        break;
      case GeometryBox::ViewBox:
        ss << "view-box";
        break;
    }
  }

  return ss.str();
}
#endif

#ifdef RN_SERIALIZABLE_STATE
folly::dynamic ClipPath::toDynamic() const {
  folly::dynamic result = folly::dynamic::object();

  if (shape) {
    result["shape"] =
        std::visit([](const auto& s) { return s.toDynamic(); }, *shape);
  }

  if (geometryBox) {
    switch (*geometryBox) {
      case GeometryBox::MarginBox:
        result["geometryBox"] = "margin-box";
        break;
      case GeometryBox::BorderBox:
        result["geometryBox"] = "border-box";
        break;
      case GeometryBox::ContentBox:
        result["geometryBox"] = "content-box";
        break;
      case GeometryBox::PaddingBox:
        result["geometryBox"] = "padding-box";
        break;
      case GeometryBox::FillBox:
        result["geometryBox"] = "fill-box";
        break;
      case GeometryBox::StrokeBox:
        result["geometryBox"] = "stroke-box";
        break;
      case GeometryBox::ViewBox:
        result["geometryBox"] = "view-box";
        break;
    }
  }

  return result;
}
#endif

} // namespace facebook::react

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/style/ClipPath.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react.uimanager.style

import android.content.Context
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.ReadableType
import com.facebook.react.uimanager.PixelUtil

/** Represents a value with a unit type (pixel, percent, etc.) */
public data class ValueUnit(
    val value: Float,
    val unit: UnitType,
) {
  public companion object {
    public fun parse(map: ReadableMap, key: String): ValueUnit? {
      if (!map.hasKey(key)) return null
      val valueMap = map.getMap(key) ?: return null
      val value = valueMap.getDouble("value").toFloat()
      val unit = valueMap.getString("unit") ?: return null
      return ValueUnit(value, UnitType.fromString(unit))
    }
  }
}

/** Unit types for values */
public enum class UnitType {
  Point,
  Percent;

  public companion object {
    public fun fromString(value: String): UnitType {
      return when (value.lowercase()) {
        "point" -> Point
        "percent" -> Percent
        else -> Point
      }
    }
  }
}

/** Circle shape: circle(radius at cx cy) */
public data class CircleShape(
    val r: ValueUnit,
    val cx: ValueUnit? = null,
    val cy: ValueUnit? = null,
) {
  public companion object {
    public fun parse(map: ReadableMap): CircleShape? {
      val r = ValueUnit.parse(map, "r") ?: return null
      val cx = ValueUnit.parse(map, "cx")
      val cy = ValueUnit.parse(map, "cy")
      return CircleShape(r, cx, cy)
    }
  }
}

/** Ellipse shape: ellipse(rx ry at cx cy) */
public data class EllipseShape(
    val rx: ValueUnit,
    val ry: ValueUnit,
    val cx: ValueUnit? = null,
    val cy: ValueUnit? = null,
) {
  public companion object {
    public fun parse(map: ReadableMap): EllipseShape? {
      val rx = ValueUnit.parse(map, "rx") ?: return null
      val ry = ValueUnit.parse(map, "ry") ?: return null
      val cx = ValueUnit.parse(map, "cx")
      val cy = ValueUnit.parse(map, "cy")
      return EllipseShape(rx, ry, cx, cy)
    }
  }
}

/** Inset shape: inset(top right bottom left round radius) */
public data class InsetShape(
    val top: ValueUnit,
    val right: ValueUnit,
    val bottom: ValueUnit,
    val left: ValueUnit,
    val borderRadius: ValueUnit? = null,
) {
  public companion object {
    public fun parse(map: ReadableMap): InsetShape? {
      val top = ValueUnit.parse(map, "top") ?: return null
      val right = ValueUnit.parse(map, "right") ?: return null
      val bottom = ValueUnit.parse(map, "bottom") ?: return null
      val left = ValueUnit.parse(map, "left") ?: return null
      val borderRadius = ValueUnit.parse(map, "borderRadius")
      return InsetShape(top, right, bottom, left, borderRadius)
    }
  }
}

/** Fill rule for polygon */
public enum class FillRule {
  NonZero,
  EvenOdd;

  public companion object {
    public fun fromString(value: String): FillRule {
      return when (value.lowercase()) {
        "nonzero" -> NonZero
        "evenodd" -> EvenOdd
        else -> NonZero
      }
    }
  }
}

/** Polygon shape: polygon(fillRule, x1 y1, x2 y2, ...) */
public data class PolygonShape(
    val points: List<Pair<ValueUnit, ValueUnit>>,
    val fillRule: FillRule? = null,
) {
  public companion object {
    public fun parse(map: ReadableMap): PolygonShape? {
      if (!map.hasKey("points")) return null
      val pointsArray = map.getArray("points") ?: return null
      val points = mutableListOf<Pair<ValueUnit, ValueUnit>>()

      for (i in 0 until pointsArray.size()) {
        val pointMap = pointsArray.getMap(i) ?: continue
        val x = ValueUnit.parse(pointMap, "x") ?: continue
        val y = ValueUnit.parse(pointMap, "y") ?: continue
        points.add(Pair(x, y))
      }

      val fillRule =
          if (map.hasKey("fillRule")) {
            FillRule.fromString(map.getString("fillRule") ?: "nonzero")
          } else {
            null
          }

      return PolygonShape(points, fillRule)
    }
  }
}

/** Rect shape: rect(top right bottom left round radius) */
public data class RectShape(
    val top: ValueUnit,
    val right: ValueUnit,
    val bottom: ValueUnit,
    val left: ValueUnit,
    val borderRadius: ValueUnit? = null,
) {
  public companion object {
    public fun parse(map: ReadableMap): RectShape? {
      val top = ValueUnit.parse(map, "top") ?: return null
      val right = ValueUnit.parse(map, "right") ?: return null
      val bottom = ValueUnit.parse(map, "bottom") ?: return null
      val left = ValueUnit.parse(map, "left") ?: return null
      val borderRadius = ValueUnit.parse(map, "borderRadius")
      return RectShape(top, right, bottom, left, borderRadius)
    }
  }
}

/** Xywh shape: xywh(x y width height round radius) */
public data class XywhShape(
    val x: ValueUnit,
    val y: ValueUnit,
    val width: ValueUnit,
    val height: ValueUnit,
    val borderRadius: ValueUnit? = null,
) {
  public companion object {
    public fun parse(map: ReadableMap): XywhShape? {
      val x = ValueUnit.parse(map, "x") ?: return null
      val y = ValueUnit.parse(map, "y") ?: return null
      val width = ValueUnit.parse(map, "width") ?: return null
      val height = ValueUnit.parse(map, "height") ?: return null
      val borderRadius = ValueUnit.parse(map, "borderRadius")
      return XywhShape(x, y, width, height, borderRadius)
    }
  }
}

/** Basic shape variant */
public sealed class BasicShape {
  public data class Circle(val shape: CircleShape) : BasicShape()

  public data class Ellipse(val shape: EllipseShape) : BasicShape()

  public data class Inset(val shape: InsetShape) : BasicShape()

  public data class Polygon(val shape: PolygonShape) : BasicShape()

  public data class Rect(val shape: RectShape) : BasicShape()

  public data class Xywh(val shape: XywhShape) : BasicShape()

  public companion object {
    public fun parse(map: ReadableMap): BasicShape? {
      if (!map.hasKey("type")) return null
      val type = map.getString("type") ?: return null

      return when (type.lowercase()) {
        "circle" -> {
          val circle = CircleShape.parse(map) ?: return null
          Circle(circle)
        }
        "ellipse" -> {
          val ellipse = EllipseShape.parse(map) ?: return null
          Ellipse(ellipse)
        }
        "inset" -> {
          val inset = InsetShape.parse(map) ?: return null
          Inset(inset)
        }
        "polygon" -> {
          val polygon = PolygonShape.parse(map) ?: return null
          Polygon(polygon)
        }
        "rect" -> {
          val rect = RectShape.parse(map) ?: return null
          Rect(rect)
        }
        "xywh" -> {
          val xywh = XywhShape.parse(map) ?: return null
          Xywh(xywh)
        }
        else -> null
      }
    }
  }
}

/** Geometry box types */
public enum class GeometryBox {
  MarginBox,
  BorderBox,
  ContentBox,
  PaddingBox,
  FillBox,
  StrokeBox,
  ViewBox;

  public companion object {
    public fun fromString(value: String): GeometryBox {
      return when (value.lowercase()) {
        "margin-box" -> MarginBox
        "border-box" -> BorderBox
        "content-box" -> ContentBox
        "padding-box" -> PaddingBox
        "fill-box" -> FillBox
        "stroke-box" -> StrokeBox
        "view-box" -> ViewBox
        else -> BorderBox
      }
    }
  }
}

/** ClipPath property value */
public data class ClipPath(
    val shape: BasicShape? = null,
    val geometryBox: GeometryBox? = null,
) {
  public companion object {
    public fun parse(map: ReadableMap?, context: Context): ClipPath? {
      if (map == null) return null

      val shape =
          if (map.hasKey("shape") && map.getType("shape") == ReadableType.Map) {
            val shapeMap = map.getMap("shape")
            if (shapeMap != null) BasicShape.parse(shapeMap) else null
          } else {
            null
          }

      val geometryBox =
          if (map.hasKey("geometryBox")) {
            GeometryBox.fromString(map.getString("geometryBox") ?: "border-box")
          } else {
            null
          }

      return ClipPath(shape, geometryBox)
    }
  }
}
